interfaces/builtin/microstack_support.go

```diff
@@ -0,0 +1,274 @@
+/sys/bus/nd/devices/{,**/} r,
+
+# For ppc device-tree access by Libvirt.
+/proc/device-tree/ r,
+/proc/device-tree/** r,
+/sys/firmware/devicetree/** r,
+
+# "virsh console" support
+/dev/pts/* rw,
+# Used by libvirt.
+/dev/ptmx rw,
+# spice
+owner /{dev,run}/shm/spice.* rw,
+
+# Used by libvirt to create lock files for /dev/pts/<num> devices
+# when handling virsh console access requests.
+/run/lock/ r,
+/run/lock/snap.@{SNAP_INSTANCE_NAME}/ rw,
+/run/lock/snap.@{SNAP_INSTANCE_NAME}/** mrwklix,
+
+# allow connect with openGraphicsFD to work
+unix (send, receive) type=stream addr=none peer=(label=libvirtd),
+unix (send, receive) type=stream addr=none peer=(label=/snap/microstack/current/usr/sbin/libvirtd),
+
+# Allow running utility processes under the specialized AppArmor profiles.
+# These profiles will prevent utility processes escaping confinement.
+capability mac_admin,
+
+# MicroStack services such as libvirt use has a server/client design where
+# unix sockets are used for IPC.
+capability chown,
+
+# MicroStack will also use privilege separation when running utility processes
+capability setuid,
+capability setgid,
+
+# Required by Nova and Neutron (e.g. to create network namespaces), see oslo.privsep docs.
+capability sys_admin,
+# Required by Neutron and OpenvSwitch.
+capability net_admin,
+capability net_broadcast,
+# Required by Nova.
+capability dac_override,
+capability dac_read_search,
+capability fowner,
+
+# Used by libvirt to alter process capabilities via prctl.
+capability setpcap,
+# Used by libvirt to create device special files.
+capability mknod,
+# Used by libvirt to write log messages to the kernel audit log.
+capability audit_write,
+# Used by QEMU for memory locking.
+capability ipc_lock,
+
+# Used my mysql.
+capability sys_nice,
+
+# ptrace
+capability sys_ptrace,
+
+# Profiles Microstack generates have a naming scheme, restrict any profile changes to
+# those matching that scheme. Need unsafe to prevent env scrubbing.
+change_profile unsafe /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/usr/bin/* -> microstack.*,
+signal (send) peer=microstack.*,
+ptrace (read, trace) peer=microstack.*,
+
+network inet stream,
+network inet6 stream,
+
+# Used by neutron-ovn-agent.
+/run/netns/ r,
+unmount /run/netns/ovnmeta-*,
+`
+
+const microStackSupportConnectedPlugSecComp = `
+# Description: allow MicroStack to operate by allowing the necessary system calls to be used by various services.
+# (libvirt, qemu, qemu-img, dnsmasq, Nova, Neutron, Keystone, Glance, Cinder)
+
+# Note that this profile necessarily contains the union of all the syscalls each of the
+# utilities requires. We rely on MicroStack to generate specific AppArmor profiles
+# for each child process, to further restrict their abilities.
+
+chown
+chown32
+fchown
+fchown32
+fchownat
+lchown
+lchownat
+`
+var microStackConnectedPlugUDev = []string{
+	`KERNEL=="vhost-net"`,
+	`KERNEL=="vhost-scsi"`,
+	`KERNEL=="vhost-vsock"`,
+	`KERNEL=="tun"`,
```

Comment:
tun is already done in network-control. I believe you need corresponding entries for vfio and nbd. I'm not sure what you want for both OTOH. I suggest creating the devices you want libvirt to have access to then using `udevadm info /dev/<thing>` and adjusting this accordingly.